src/places/autocomplete_service.ts

```diff
@@ -1,32 +1,82 @@
-}
-interface AutocompletePrediction {
-  description: string;
-  place_id: string;
-}
-interface QueryAutocompletePrediction {
-  description: string;
-  matched_substrings: google.maps.places.PredictionSubstring[];
-  place_id?: string;
-  reference?: string;
-  terms: google.maps.places.PredictionTerm[];
-}
+// Handle location/bounds restrictions. bounds and location have been deprecated, and in some cases
+// have actually been removed (although still mentioned in the documentation as only deprecated).
+//   * locationBias is the top preferred field, and can be MigrationLatLng|LatLngLiteral|MigrationLatLngBounds|LatLngBoundsLiteral
+//   * bounds / locationRestriction is the next preferred field
+//   * location is the final field that is checked
+//   * radius must be paired with a LatLng (locationBias / location) to specify a circle bias
+// This logic can be shared between both getQueryPredictions and getPlacePredictions
+const setRequestLocationBias = (
+  input: AutocompleteRequest | SuggestRequest,
+  location,
+  locationBias,
+  bounds,
+  radius,
+) => {
+  let inputBounds, inputLocation;
+  if (locationBias) {
+    // MigrationLatLng|LatLngLiteral
+    if (locationBias.lat !== undefined && locationBias.lng !== undefined) {
+      inputLocation = new MigrationLatLng(locationBias);
+    } /* MigrationLatLngBounds|LatLngBoundsLiteral */ else {
+      inputBounds = new MigrationLatLngBounds(locationBias);
+    }
+  } else if (bounds) {
+    inputBounds = new MigrationLatLngBounds(bounds);
+  } else if (location) {
+    inputLocation = new MigrationLatLng(location);
+  }
+
+  // If bounds was found, then location is ignored
+  if (inputBounds) {
+    const southWest = inputBounds.getSouthWest();
+    const northEast = inputBounds.getNorthEast();
 
+    input.Filter = {
+      BoundingBox: [southWest.lng(), southWest.lat(), northEast.lng(), northEast.lat()],
+    };
+  } else if (inputLocation) {
+    // If we have a location and a radius, then we will use a circle
+    // Otherwise, just the location will be used
+    const lngLat = LatLngToLngLat(inputLocation);
+    if (lngLat) {
+      if (radius) {
+        input.Filter = {
+          Circle: {
+            Center: lngLat,
+            Radius: radius,
+          },
+        };
+      } else {
+        input.BiasPosition = lngLat;
+      }
+    }
+  }
+};
+
+// Migration class for google.maps.places.AutocompleteService
+// https://developers.google.com/maps/documentation/javascript/reference/places-autocomplete-service#AutocompleteService
 export class MigrationAutocompleteService {
   _client: GeoPlacesClient; // This will be populated by the top level module that creates our location client
 
-  getQueryPredictions(request, callback: (a: QueryAutocompletePrediction[], b: PlacesServiceStatus) => void) {
+  // https://developers.google.com/maps/documentation/javascript/reference/places-autocomplete-service#AutocompleteService.getQueryPredictions
+  getQueryPredictions(
+    request,
```

Comment:
request doesn't have a Type, but we assume specific properties on it

Reply:
The answer to this goes in hand with the deprecation message above, so I will give a brief synopsis here as well :)

This request type that we would get from `@types/google.maps` doesn't match Google's implementation. In there documentation, there are several fields marked as deprecated that are actually removed from the published types, but you can actually still call their API with them and they are respected.

I can add our own type for it with a comment giving context.

Reply:
Okay, that makes sense.  We can't really list a stable type if that type ultimately isn't ours and we know it's in some phase of deprecation.